src/main/cpp/MpvAvPlayback.cpp

```diff
@@ -0,0 +1,527 @@
+	return MPV_ERROR_SUCCESS;
+}
+
+double MpvAvPlayback::GetRate()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	double currentRate;
+	int err = _mpvGetProperty(_mpvHandle, "speed", MPV_FORMAT_DOUBLE, &currentRate);
+	if (err < 0) {
+		return err;
+	}
+
+	return currentRate;
+}
+
+float MpvAvPlayback::GetFps()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	if (!_streamFps || _streamFps <= 0 || _streamFps == NULL) {
+		int err = _mpvGetProperty(_mpvHandle, "container-fps", MPV_FORMAT_DOUBLE, &_streamFps);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return _streamFps;
+}
+
+int MpvAvPlayback::GetImageWidth()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	if (!_imageWidth || _imageWidth <= 0 || _imageWidth == NULL) {
+		int err = _mpvGetProperty(_mpvHandle, "width", MPV_FORMAT_INT64, &_imageWidth);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return _imageWidth;
+}
+
+int MpvAvPlayback::GetImageHeight()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	if (!_imageHeight|| _imageHeight <= 0 || _imageHeight == NULL) {
+		int err = _mpvGetProperty(_mpvHandle, "height", MPV_FORMAT_INT64, &_imageHeight);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return _imageHeight;
+}
+
+double MpvAvPlayback::GetDuration()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	if (!_streamDuration || _streamDuration <= 0 || _streamDuration == NULL) {
+		// TODO(Reda) Remove this, it is just for testing, need to go through async calls
+		int err = _mpvGetProperty(_mpvHandle, "duration", MPV_FORMAT_DOUBLE, &_streamDuration);
+		if (err < 0) {
+			return err;
+		}
+	}
+
+	return _streamDuration;
+}
+
+int MpvAvPlayback::StepBackward()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	const char *cmd[] = { "frame-back-step", NULL, NULL };
+
+	int err = DoMpvCommand(cmd);
+	if (err < 0) {
+		return err;
+	}
+
+	return MPV_ERROR_SUCCESS;
+}
+
+int MpvAvPlayback::StepForward()
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	const char *cmd[] = { "frame-step", NULL, NULL };
+
+	int err = DoMpvCommand(cmd);
+	if (err < 0) {
+		return err;
+	}
+
+	return MPV_ERROR_SUCCESS;
+}
+
+int MpvAvPlayback::DoMpvCommand(const char **cmd)
+{
+	if (!_mpvHandle) {
+		return MPV_ERROR_GENERIC;
+	}
+
+	int err = _mpvCommand(_mpvHandle, cmd);
+	if (err < 0) {
+		printf("Cannot execute the command");
+		return MPV_ERROR_COMMAND;
+	}
+
+	return MPV_ERROR_SUCCESS;
+}
+
+int MpvAvPlayback::Destroy()
+{
+	int err = 0;
+	err = _mpvTerminateDestroy(_mpvHandle);
+	if (err < 0) {
+		return err;
+	}
+
+	return MPV_ERROR_SUCCESS;
+}
+
+void MpvAvPlayback::init_and_event_loop(const char *filename)
+{
+	if (SDL_Init(SDL_INIT_VIDEO) < 0) {
+		printf("SDL init failed");
+		exit(1);
+	}
+
+	SDL_Window *window =
+		SDL_CreateWindow(filename, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
+			1000, 500, SDL_WINDOW_SHOWN |
+			SDL_WINDOW_RESIZABLE);
+
+	if (!window) {
+		printf("failed to create SDL window");
+		exit(1);
```

Comment:
Q: Do we need to cleanup some memory before doing this hard exit?
A: No, there is no need for a cleanup, since the player will be initialized after making sure that a window is available.
Q: Can we avoid this hard exit?
A: No, the MPV player need a window ID in order to properly work.